url_shortener/views/handlers.py

```diff
@@ -1,60 +1,332 @@
+    else:
+        return Response(status=httplib.CONFLICT, json_body={
+            'status': 'error',
+            'description': 'Smth went wrong.'
+        })
 
-def protected_resource_write_example(request: Request) -> Response:
-    key = request.matchdict['key']
-    value = request.json_body.get('value')
 
-    if value is None:
-        return Response(status=httplib.BAD_REQUEST,
-                        json_body={
-                            'status': 'error',
-                            'reason':
-                            '`value` was not provided within request',
-                        })
+def redirect_longlink(request: Request):
+    """ Redirects if conditions are met.
+        CONDITIONS:
+            - If data entered is non-null
+            - If key exists.
+            - If key is valid.
+        Returns HTTPFound
+    """
 
     logic: Logic = request.registry.logic
-    success = logic.save_example_if_not_exists(key, value)
+    try:
+        id = request.matchdict['id']
+    except Exception as e:
+        print(e)
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Request body is missing.'
+        })
+
+    try:
+        long_link = logic.read_by_key(id)
+    except Exception as e:
+        print(e)
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Key does not exist.'
+        })
+
+    if id is None or long_link is None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Incorrect key.'
+        })
 
+    return HTTPFound(location=long_link)
+
+
+def create_shortlink(request: Request) -> Response:
+    """ Creates a shortlink id, by using id if provided or generating 4 char random string."""
+    logic: Logic = request.registry.logic
+
+    try:
+        url_provided = request.json_body.get('url')
+        id_provided = request.json_body.get('id')
+    except Exception as e:
+        print(e)
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Request body is missing.'
+        })
+
+    if url_provided is None:
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'URL must be provided.'
+        })
+
+    # If custom id is not provided, or this id already exists, we give an random one.
+    while id_provided is None or logic.read_by_key(id_provided) is not None:
+        id_provided = ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase) for _ in range(4))
+
+    # People shouldn't overwrite user id's so we'll block it.
+    if any(c.isdigit() for c in id_provided):
+        return Response(status=httplib.BAD_REQUEST, json_body={
+            'status': 'error',
+            'description': 'Numbers in id is not allowed.'
```

Comment:
But why numbers are not allowed? What if I want to have shortened url like `https://localhost:6543/lab_1` ?

Reply:
We store users in key: value fashion. Did not wanted user to overwrite on users. 

Improved this by adding a condition of id should contain at least one alphabetic char.